atecc608a_se.c

```diff
@@ -0,0 +1,227 @@
+/**
+ * \file atecc608a_se.c
+ * \brief Secure element driver implementation for ATECC508A and ATECC509A.
+ */
+
+/*
+ *  Copyright (C) 2019, ARM Limited, All Rights Reserved
+ *  SPDX-License-Identifier: Apache-2.0
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License"); you may
+ *  not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ *  WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+#include "atecc608a_se.h"
+#include "atca_helpers.h"
+
+#ifdef DEBUG_PRINT
+#include <stdio.h>
+#endif
+
+#include <stdbool.h>
+#include <stdint.h>
+
+/* Uncomment to print results on success */
+//#define DEBUG_PRINT
+
+/** This macro checks if the result of an `expression` is equal to an
+ *  `expected` value and sets a `status` variable of type `psa_status_t` to
+ *  `PSA_SUCCESS`. If they are not equal, the `status` is set to
+ *  `psa_error instead`, and the code jumps to the `exit` label. */
+#define ASSERT_STATUS(expression, expected, psa_error)          \
+    do                                                          \
+    {                                                           \
+        ATCA_STATUS ASSERT_result = (expression);               \
+        ATCA_STATUS ASSERT_expected = (expected);               \
+        if ((ASSERT_result) != (ASSERT_expected))               \
+        {                                                       \
+            status = (psa_error);                               \
+            goto exit;                                          \
+        }                                                       \
+        status = PSA_SUCCESS;                                   \
+    } while(0)
+
+/** Check if an ATCA operation is successful, translate the error otherwise. */
+#define ASSERT_SUCCESS(expression) ASSERT_STATUS(expression, ATCA_SUCCESS, \
+                                      atecc608a_to_psa_error(ASSERT_result))
+
+ATCAIfaceCfg atca_iface_config = {
+    .iface_type = ATCA_I2C_IFACE,
+    .devtype = ATECC608A,
+    .atcai2c.slave_address = 0xC0,
+    .atcai2c.bus = 2,
+    .atcai2c.baud = 400000,
+    .wake_delay = 1500,
+    .rx_retries = 20,
+};
+
+psa_status_t atecc608a_to_psa_error(ATCA_STATUS ret)
```

Comment:
What use this this outside the driver? Why shouldn't this be static?

If an application is working with PSA error types, I'd expect it to always use those types and not be fiddling with ATCA statuses. A PSA Service might find this useful, potentially.

Reply:
This can be used by all other utility functions that we write for this secure element that do not fit the driver - like provisioning, config printing, misc. operations.